black8/lint.py

```diff
@@ -0,0 +1,167 @@
+import fnmatch
+import shutil
+import subprocess
+import sys
+from contextlib import contextmanager
+from pathlib import Path
+from subprocess import PIPE, CalledProcessError
+
+import click
+
+from . import ASSETS_DIR, cli
+
+# -----------------------------------------------------------------------------
+
+CONFIG_FILE = Path(".black8")
+
+SNAPSHOT_GLOB = "*/snapshots/snap_*.py"
+SNAPSHOT_REGEX = r".*\/snapshots\/snap_.*\.py"
+CONFIGURATION_FILES = (".flake8", ".isort.cfg", "pyproject.toml")
+
+# -----------------------------------------------------------------------------
+
+
+def get_project_paths():
+    return CONFIG_FILE.read_text().split()
+
+
+def get_dirty_filenames(paths, *, staged=False):
+    filenames = subprocess.run(
+        (
+            "git",
+            "diff-index",
+            "--name-only",
+            "--diff-filter",
+            "ACM",
+            *(("--cached",) if staged else ()),
+            "HEAD",
+            "--",
+            *paths,
+        ),
+        check=True,
+        stdout=PIPE,
+        encoding=sys.getdefaultencoding(),
+    ).stdout.split()
+
+    return tuple(
+        filename
+        for filename in filenames
+        if Path(filename).suffix == ".py"
+        and not fnmatch.fnmatch(filename, SNAPSHOT_GLOB)
+    )
+
+
+# -----------------------------------------------------------------------------
+
+
+def copy_configuration(override=False):
+    for name in CONFIGURATION_FILES:
+        if override is True or not Path(name).exists():
+            # This will always be in project root because we implicitly assert
+            # in get_project_paths().
+            shutil.copy(ASSETS_DIR / name, ".")
+
+
+def black(paths, *, check=False):
+    subprocess.run(
+        (
+            "black",
+            *(("--check", "--diff") if check else ()),
+            "--exclude",
+            SNAPSHOT_REGEX,
+            "--quiet",
+            "--",
+            *paths,
+        ),
+        check=True,
+    )
+
+
+def isort(paths, *, project_paths, check=False):
+    subprocess.run(
+        (
+            "isort",
+            *(("--check", "--diff") if check else ("--apply",)),
+            *(arg for path in project_paths for arg in ("--project", path)),
+            "--skip-glob",
+            SNAPSHOT_GLOB,
+            "--atomic",
+            "--quiet",
+            "--recursive",
+            "--",
+            *paths,
+        ),
+        check=True,
+    )
+
+
+def flake8(paths, *, check=True):
+    assert check is True, "Flake8 has no format mode"
+
+    subprocess.run(("flake8", "--", *paths), check=True)
+
+
+# -----------------------------------------------------------------------------
+
+
+@cli.command(
+    help=f"check code style. If no file is specified, it will run on all the files specified in {CONFIG_FILE}"
+)
+@click.option("-c", "--override-config", is_flag=True)
+@click.argument("files", nargs=-1)
+def lint(override_config, files):
```

Comment:
we could actually get rid of a lint function completely and solely rely on pre-commit, but I think it doesn't make sense to strictly depend on pre-commit